src/backend/distributed/planner/distributed_planner.c

```diff
@@ -593,6 +566,253 @@ CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query
+	 * plan the query as a router query. If not supported, otherwise try
+	 * the full blown plan/optimize/physical planing process needed to
+	 * produce distributed query plans.
+	 */
+	if (EnableRouterExecution)
+	{
+		RelationRestrictionContext *relationRestrictionContext =
+			plannerRestrictionContext->relationRestrictionContext;
+
+		distributedPlan = CreateRouterPlan(originalQuery, query,
+										   relationRestrictionContext);
+		if (distributedPlan != NULL)
+		{
+			if (distributedPlan->planningError == NULL)
+			{
+				/* successfully created a router plan */
+				return distributedPlan;
+			}
+			else
+			{
+				/*
+				 * For debugging it's useful to display why query was not
+				 * router plannable.
+				 */
+				RaiseDeferredError(distributedPlan->planningError, DEBUG1);
+			}
+		}
+	}
+
+	if (hasUnresolvedParams)
+	{
+		/*
+		 * There are parameters that don't have a value in boundParams.
+		 *
+		 * The remainder of the planning logic cannot handle unbound
+		 * parameters. We return a NULL plan, which will have an
+		 * extremely high cost, such that postgres will replan with
+		 * bound parameters.
+		 */
+		return NULL;
+	}
+
+	/*
+	 * If there are parameters that do have a value in boundParams, replace
+	 * them in the original query. This allows us to more easily cut the
+	 * query into pieces (during recursive planning) or deparse parts of
+	 * the query (during subquery pushdown planning).
+	 */
+	originalQuery = (Query *) ResolveExternalParams((Node *) originalQuery,
+													boundParams);
+
+	/*
+	 * Plan subqueries and CTEs that cannot be pushed down by recursively
+	 * calling the planner and add the resulting plans to subPlanList.
+	 */
+	error = RecursivelyPlanSubqueriesAndCTEs(originalQuery, plannerRestrictionContext,
+											 planId, &subPlanList);
+	if (error != NULL)
+	{
+		RaiseDeferredError(error, ERROR);
+	}
+
+	/*
+	 * If subqueries were recursively planned then we need to replan the query
+	 * to get the new planner restriction context and apply planner transformations.
+	 *
+	 * We could simplify this code if the logical planner was capable of dealing
+	 * with an original query. In that case, we would only have to filter the
+	 * planner restriction context.
+	 */
+	if (list_length(subPlanList) > 0)
+	{
+		Query *newQuery = copyObject(originalQuery);
+		bool setPartitionedTablesInherited = false;
+
+		/* remove the pre-transformation planner restrictions context */
+		PopPlannerRestrictionContext();
+
+		/* create a fresh new planner context */
+		plannerRestrictionContext = CreateAndPushPlannerRestrictionContext();
+
+		/*
+		 * We force standard_planner to treat partitioned tables as regular tables
+		 * by clearing the inh flag on RTEs. We already did this at the start of
+		 * distributed_planner, but on a copy of the original query, so we need
+		 * to do it again here.
+		 */
+		AdjustPartitioningForDistributedPlanning(newQuery, setPartitionedTablesInherited);
```

Comment:
Could you add the following test since it is fixed by this call:
```SQL
WITH cte AS
  ( SELECT DISTINCT id
   FROM partitioning_test)
SELECT *
FROM cte,

  (SELECT *,
          random()
   FROM partitioning_test) AS foo
WHERE cte.id = foo.id;

```

Reply:
Added Furkan's partitioning tests.

Reply:
Those tests don't exercise the case that final query includes a partitioned relation. So, adding such a test might be valuable.

Reply:
Seems like they do:
https://github.com/citusdata/citus/pull/1853/files#diff-2bbbc00d8a6593399a358da881bacf83R84

Reply:
Ah, and the query needs to go through subquery pushdown planning (or potentially router planning as well) since the tricky part is that restriction information includes the partition's restriction information unless this line is added.

Reply:
CTEs always go through subquery pushdown planning if not a router query.

Reply:
That's true. But, the recursive call to `distributed_planner` already handles partitioning for the subqueries pointed by the `CTE`s. However, the final query might go from subquery pushdown planner as well, and those tests don't exercise such case.

Reply:
(and read_intermediate_result is volatile, so never flattened)

Reply:
> However, the final query might go from subquery pushdown planner as well, and those tests don't exercise such case.

This one does:
https://github.com/citusdata/citus/pull/1853/files#diff-2bbbc00d8a6593399a358da881bacf83R84